### PR TITLE
ECO-967: Feature/store order

### PR DESCRIPTION
**Pull request by ghost**
- Developer(s): @sergeysikachev

- Peer Reviewer:

- Ticket: URL_HERE

- Project PR: URL_HERE

- Academy PR: URL_HERE (for feature documentation and if needed migration guides and project integration guide)

#### Please confirm

- [ ] No new OS components - or they have been approved by legal department
- [ ] All new or heavily changed classes get an "A" rating (Scrutinizer), except Facades, Factories, QueryContainers and Tests

#### Documentation

- [ ] Functional documentation provided or in progress?
- [ ] Integration guide for projects provided or not needed?
- [ ] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraints         |
   :--------------------- | :------------------------ | :--------------------- |
   ArvatoRss              |       major                 |                       |

#### Release Notes

- Store Order functionality added.
- Risk Check functionality partially refactored during store order implementation.
-----------------------------------------

#### Bundle ArvatoRss

_**Major:** Backwards-compatible feature_

Def of done (by responsible developer):
- [ ] All changes are backward-compatible. Outdated code is marked as deprecated
- [ ] The change is isolated and not mixed with any cleanups or other changes
- [ ] New code fits to the architecture rules
- [ ] All new or changed business logic is covered by functional tests (in Zed business layer)

##### Change log

- Store Order functionality added.
- Risk Check functionality partially refactored during store order implementation.

### CODE REVIEW

File: src/SprykerEco/Zed/ArvatoRss/Business/Api/Adapter/SoapApiAdapter.php [hunk 1]
```diff
@@ -73,17 +61,44 @@ public function performRiskCheck(ArvatoRssRiskCheckRequestTransfer $requestTrans
...
+        );
+
+        $result = $soapClient->StoreOrder($params);
+        try {
+            $this->validateResponse($result);
+        } catch (ArvatoRssStoreOrderApiException $exception) {
+            $responseTransfer = new ArvatoRssStoreOrderResponseTransfer();
+            $responseTransfer->setIsError(true);
+            $responseTransfer->setErrorMessage($exception->getMessage());
+            return $responseTransfer;
```
Comment (alex-galych) [inline src/SprykerEco/Zed/ArvatoRss/Business/Api/Adapter/SoapApiAdapter.php]: Empty line before return is missing.

File: src/SprykerEco/Zed/ArvatoRss/Business/Api/Converter/StoreOrderResponseConverterInterface.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+use stdClass;
+
+interface StoreOrderResponseConverterInterface
+{
+    /**
+     * @param \stdClass $response
+     *
+     * @return \Generated\Shared\Transfer\ArvatoRssStoreOrderResponseTransfer
+     */
+    public function convert(stdClass $response);
```
Comment (alex-galych) [inline src/SprykerEco/Zed/ArvatoRss/Business/Api/Converter/StoreOrderResponseConverterInterface.php]: You can use \stdClass without use statement for your class

**Review by alex-galych — APPROVED** [review]

File: tests/SprykerEcoTest/Zed/ArvatoRss/_support/Mock/MoneyFacadeMock.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace SprykerEcoTest\Zed\ArvatoRss\Mock;
+
+use SprykerEco\Zed\ArvatoRss\Dependency\Facade\ArvatoRssToMoneyInterface;
+
+class MoneyFacadeMock implements ArvatoRssToMoneyInterface
+{
+    const DECIMAL_VALUE = 100.00;
```
Comment (dereuromark) [inline tests/SprykerEcoTest/Zed/ArvatoRss/_support/Mock/MoneyFacadeMock.php]: Constants usually one groups by the common part: VALUE_X vs VALUE_X etc.

File: tests/SprykerEcoTest/Zed/ArvatoRss/Business/Facade/RiskCheckFacadeTest.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            $response->getArvatoRssQuoteData()
+                ->getArvatoRssRiskCheckResponse()
+        );
+
+        $riskCheckResponse = $response->getArvatoRssQuoteData()->getArvatoRssRiskCheckResponse();
+        $this->assertEquals(static::RESPONSE_STRING_FIElD_VALUE, $riskCheckResponse->getResult());
+        $this->assertEquals(static::RESPONSE_STRING_FIElD_VALUE, $riskCheckResponse->getErrorMessage());
+        $this->assertEquals(static::RESPONSE_STRING_FIElD_VALUE, $riskCheckResponse->getResultCode());
+        $this->assertEquals(static::RESPONSE_STRING_FIElD_VALUE, $riskCheckResponse->getActionCode());
+        $this->assertEquals(static::IS_ERROR, $riskCheckResponse->getIsError());
```
Comment (dereuromark) [inline tests/SprykerEcoTest/Zed/ArvatoRss/Business/Facade/RiskCheckFacadeTest.php]: That sounds like a code smell: Please dont const trivial things like false here, and instead use

    $this->assertFalse($...)

File: tests/SprykerEcoTest/Zed/ArvatoRss/Business/Facade/AbstractFacadeTest.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+use SprykerEco\Shared\ArvatoRss\ArvatoRssConstants;
+use SprykerEcoTest\Zed\ArvatoRss\Helper\QuoteHelper;
+use SprykerTest\Shared\Testify\Helper\ConfigHelper;
+
+class AbstractFacadeTest extends Test
+{
+    const IS_ERROR = false;
+    const RESPONSE_STRING_FIElD_VALUE = 'test';
+
+    protected $quote;
```
Comment (dereuromark) [inline tests/SprykerEcoTest/Zed/ArvatoRss/Business/Facade/AbstractFacadeTest.php]: seems to miss a doc block about its type.

File: tests/SprykerEcoTest/Zed/ArvatoRss/Business/Api/Mapper/Aspect/OrderMapperTest.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+use Generated\Shared\DataBuilder\StoreBuilder;
+use Generated\Shared\Transfer\ArvatoRssOrderTransfer;
+use SprykerEco\Zed\ArvatoRss\Business\Api\Mapper\Aspect\OrderMapper;
+use SprykerEco\Zed\ArvatoRss\Dependency\Facade\ArvatoRssToMoneyInterface;
+use SprykerEco\Zed\ArvatoRss\Dependency\Facade\ArvatoRssToStoreInterface;
+
+class OrderMapperTest extends AbstractMapperTest
+{
+    /**
+     * @const double DECIMAL_VALUE
```
Comment (dereuromark) [inline tests/SprykerEcoTest/Zed/ArvatoRss/Business/Api/Mapper/Aspect/OrderMapperTest.php]: float - and no repeating of the name DECIMAL_VALUE required

File: src/SprykerEco/Zed/ArvatoRss/Business/Api/Mapper/Aspect/BillingCustomerMapper.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+use Generated\Shared\Transfer\AddressTransfer;
+use Generated\Shared\Transfer\ArvatoRssBillingCustomerTransfer;
+use Generated\Shared\Transfer\ArvatoRssCustomerAddressTransfer;
+use Generated\Shared\Transfer\QuoteTransfer;
+use SprykerEco\Service\ArvatoRss\Iso3166ConverterServiceInterface;
+
+class BillingCustomerMapper implements BillingCustomerMapperInterface
+{
+    /**
+     * @const string DATE_FORMAT
```
Comment (dereuromark) [inline src/SprykerEco/Zed/ArvatoRss/Business/Api/Mapper/Aspect/BillingCustomerMapper.php]: Retype DATE_FORMAT is not needed here

File: src/SprykerEco/Zed/ArvatoRss/Business/Api/Mapper/Aspect/BillingCustomerMapper.php [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    const DATE_FORMAT = 'Y-m-d';
+
+    /**
+     * @var \SprykerEco\Service\ArvatoRss\Iso3166ConverterServiceInterface $iso3166
```
Comment (dereuromark) [inline src/SprykerEco/Zed/ArvatoRss/Business/Api/Mapper/Aspect/BillingCustomerMapper.php]: $iso3166 is wrong and should be removed

File: src/SprykerEco/Zed/ArvatoRss/Business/Api/Converter/StoreOrderResponseConverter.php [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+use stdClass;
+
+class StoreOrderResponseConverter implements StoreOrderResponseConverterInterface
+{
+    /**
+     * @param \stdClass $response
+     *
+     * @return \Generated\Shared\Transfer\ArvatoRssStoreOrderResponseTransfer
+     */
+    public function convert(stdClass $response)
```
Comment (dereuromark) [inline src/SprykerEco/Zed/ArvatoRss/Business/Api/Converter/StoreOrderResponseConverter.php]: Why stdClass and not a more concrete implementation or interface?

File: src/SprykerEco/Zed/ArvatoRss/ArvatoRssConfig.php [hunk 10]
```diff
@@ -27,4 +27,16 @@ public function getAuthorization()
     {
         return $this->get(ArvatoRssConstants::ARVATORSS_AUTHORISATION);
     }
+
+    /**
+     * @param string $paymentMethod
+     *
+     * @return mixed
```
Comment (dereuromark) [inline src/SprykerEco/Zed/ArvatoRss/ArvatoRssConfig.php]: really mixed?

File: src/SprykerEco/Zed/ArvatoRss/Business/Api/Mapper/Aspect/BillingCustomerMapper.php (hunk 8 above)
Comment (sergeysikachev) [reply src/SprykerEco/Zed/ArvatoRss/Business/Api/Mapper/Aspect/BillingCustomerMapper.php]: Could you please explain why?

File: config/config.dist.php [hunk 11]
```diff
@@ -5,3 +5,6 @@
 $config[ArvatoRssConstants::ARVATORSS_URL] = '';
 $config[ArvatoRssConstants::ARVATORSS_CLIENTID] = '';
 $config[ArvatoRssConstants::ARVATORSS_AUTHORISATION] = '';
+$config[ArvatoRssConstants::ARVATORSS_PAYMENT_TYPE_MAPPING] = [
+    'invoice' => \SprykerEco\Shared\ArvatoRss\ArvatoRssApiConfig::INVOICE,
```
Comment (dereuromark) [inline config/config.dist.php:9]: This magic string comes up a few times. Maybe it can be a class const in config too? For to be reused in tests as well.

Comment (ghost) [reply config/config.dist.php:9]: It's supposed to be a $payment->getPaymentMethod() return value. It depends on payment methods available int the system. I didn't find a constant responsible for this.
For example, I'm interested in values in $paymentMethods array here. *vendor/spryker/spryker/Bundles/DummyPayment/src/Spryker/Yves/DummyPayment/Handler/DummyPaymentHandler.php*
Probably some other approach should be used here, but for now I see it only the way I implemented.

**Pull request merged**